File: 2024/day1/day1.exs
defmodule Day1 do
  def part1(path) do
    File.stream!(path, :line)
    |> Stream.map(&String.trim/1)
    |> Stream.map(&String.split(&1, " ", [trim: true]))
    |> Enum.to_list()
    |> Enum.reduce({[], []}, fn [l, r], {acc_l, acc_r} ->
      {[String.to_integer(l) | acc_l], [String.to_integer(r) | acc_r]}
    end)
    |> then(fn {l, r} -> Enum.zip(Enum.sort(l), Enum.sort(r)) end)
    |> Enum.map(fn {l, r} -> abs(r-l) end)
    |> Enum.sum()
  end

  def part2(path) do
    File.stream!(path, :line)
    |> Stream.map(&String.trim/1)
    |> Stream.map(&String.split(&1, " ", [trim: true]))
    |> Enum.to_list()
    |> Enum.reduce({[], []}, fn [l, r], {acc_l, acc_r} ->
      {[String.to_integer(l) | acc_l], [String.to_integer(r) | acc_r]}
    end)
    |> then(fn {l, r} -> {l, Enum.frequencies(r)} end)
    |> then(fn {l, r_freq} ->
      Enum.map(l, fn value ->
        value * Map.get(r_freq, value, 0)
      end)
    end)
    |> Enum.sum()
  end
end

test1_correct = 11
test2_correct = 31
test = Day1.part1("day1/data/day1_test.txt")
if test != test1_correct do
    raise "Part 1 test failed; it returned #{test} instead of #{test1_correct}"
end
IO.puts "Part 1 answer from elixir is: #{Day1.part1("day1/data/day1.txt")}"

test = Day1.part2("day1/data/day1_test.txt")
if test != test2_correct do
    raise "Part 2 test failed; it returned #{test} instead of #{test2_correct}"
end
IO.puts "Part 2 answer from elixir is: #{Day1.part2("day1/data/day1.txt")}"

File: 2024/day1/day1.py
from collections import Counter
from pathlib import Path
from time import perf_counter

def load_input(test=False, file_path=None):
    file_path = Path(f'day3', 'data', f'{"test" if test else "input"}.txt') if not file_path else file_path
    with open(file_path, 'r') as f:
        return f.readlines()

def part1(test=False, file_path=None):
    inp = load_input(test, file_path)
    start_time = perf_counter()
    left, right = sorted([int(line.split()[0]) for line in inp]), sorted([int(line.split()[1]) for line in inp])
    return sum([abs(l - r) for l, r in zip(left, right)]), perf_counter() - start_time

def part2(test=False, file_path=None):
    inp = load_input(test, file_path)
    start_time = perf_counter()
    left, right = [int(line.split()[0]) for line in inp], Counter([int(line.split()[1]) for line in inp])
    return sum([l*right[l] for l in left]), perf_counter() - start_time

test1_correct = 11
test2_correct = 31
test, _ = part1(test=True)
assert test == test1_correct, f'Part 1 test failed; it returned {test} instead of {test1_correct}'
part1_ans, part1_time = part1()
print(f'Part 1 answer is: {part1_ans}, returned in {part1_time * 1000:.3f} ms')
test, _ = part2(test=True)
assert test == test2_correct, f'Part 2 test failed; it returned {test} instead of {test2_correct}'
part2_ans, part2_time = part2()
print(f'Part 2 answer is: {part2_ans}, returned in {part2_time * 1000:.3f} ms')